src/Stratis.Bitcoin/Utilities/AsyncQueue.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Async queue is a simple thread-safe queue that asynchronously executes 
+    /// a user-defined callback when a new item is added to the queue.
+    /// <para>
+    /// The queue guarantees that the user-defined callback is executed only once at the time. 
+    /// If another item is added to the queue, the callback is called again after the current execution 
+    /// is finished.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="T">Type of items to be inserted in the queue.</typeparam>
+    public class AsyncQueue<T>: IDisposable
+    {
+        /// <summary>
+        /// Represents a callback method to be executed when a new item is added to the queue.
+        /// </summary>
+        /// <param name="item">Newly added item.</param>
+        /// <param name="cancellationToken">Cancellation token that the callback method should use for its async operations to avoid blocking the queue during shutdown.</param>
+        public delegate Task OnEnqueueAsync(T item, CancellationToken cancellationToken);
+
+        /// <summary>Storage of items in the queue that are waiting to be consumed.</summary>
+        private readonly ConcurrentQueue<T> items;
+
+        /// <summary>Event that is triggered when at least one new item is waiting in the queue.</summary>
+        private readonly AsyncManualResetEvent signal;
+
+        /// <summary>Callback routine to be called when a new item is added to the queue.</summary>
+        private readonly OnEnqueueAsync onEnqueueAsync;
+
+        /// <summary>Consumer of the items in the queue which responsibility is to execute the user defined callback.</summary>
+        private readonly Task consumerTask;
+
+        /// <summary>Cancellation that is triggered when the component is disposed.</summary>
+        private readonly CancellationTokenSource cancellationTokenSource;
+
+        /// <summary>
+        /// Initializes the queue.
+        /// </summary>
+        /// <param name="onEnqueueAsync">Callback routine to be called when a new item is added to the queue.</param>
+        public AsyncQueue(OnEnqueueAsync onEnqueueAsync)
```

Comment:
Why not using `Func<T, CancellationToken, Task>` instead of creating a delegate?

Reply:
the ONLY reason for that is to have comments on the arguments of the delegate, 
I originally did have Func there and found out i lack the documentation of those parameters

i checked with MSDN and they do use delegates for the same reason (at least i believe), those delegates are commented separately and that's what i needed here